File: model/document/Waybill.js
/**
 * Классы и типы для работы для работы с товарно-транспортной накладной.
 * @module module:waybill
 */
import ProductInfo from '../dictionary/ProductInfo'
import {Direction, UnitType, Version, WaybillResolution, WaybillStatus, WaybillType} from "../types"

/**
 * @class module:waybill.Transport
 * @classdesc Транспорт.
 * @param {?string} type Тип перевозки
 * @param {?string} company Название компании перевозчика
 * @param {?string} car Автомобиль
 * @param {?string} trailer Прицеп
 * @param {?string} customer Заказчики
 * @param {?string} driver Водитель
 * @param {?string} loadPoint Место погрузки
 * @param {?string} unloadPoint Место разгрузки
 * @param {?string} redirect Перенаправление
 * @param {?string} forwarder Экспедитор
 */
export class Transport {
    constructor(type: (string | null),
                company: (string | null),
                car: (string | null),
                trailer: (string | null),
                customer: (string | null),
                driver: (string | null),
                loadPoint: (string | null),
                unloadPoint: (string | null),
                redirect: (string | null),
                forwarder: (string | null)) {
        this.type = type;
        this.company = company;
        this.car = car;
        this.trailer = trailer;
        this.customer = customer;
        this.driver = driver;
        this.loadPoint = loadPoint;
        this.unloadPoint = unloadPoint;
        this.redirect = redirect;
        this.forwarder = forwarder;
    }
}

/**
 * @class module:waybill.Waybill
 * @classdesc Товарно-транспортная накладная (ТТН).
 * @param {string} uuid уникальный идентификатор
 * @param {string} docOwner кто подает документы
 * @param {?string} identity Идентификатор накладной (клиентский, к заполнению необязательный)
 * @param {module:waybill#WaybillType} type Тип накладной
 * @param {?module:waybill#UnitType} unitType Тип поставки (упакованная или нет)
 * @param {string} number Номер документа
 * @param {Date} date Дата составления
 * @param {Date} shippingDate  Дата отгрузки продукции
 * @param {?module:waybill.Transport} transport Информация о транспорте
 * @param {string} shipperId Грузоотправитель
 * @param {string} consigneeId Грузополучатель
 * @param {?string} supplierId Поставщик
 * @param {?string} base Основание
 * @param {?string} note Заметки
 * @param {module:waybill#WaybillStatus} status Состояние накладной
 * @param {module:waybill#WaybillResolution} resolution резолюция по накладной
 * @param {?string} ttnInformF2RegUuid uuid формы 2 для накладной
 * @param {?string} wbRegId ИД накладной в системе (присвоенный)
 * @param {module:waybill#Direction} direction Направление документа в представлени УТМ (входящий/исходящий)
 * @param {Version} version Версия протокола ЕГАИС, по которому отправлена накладная
 */
export default class Waybill {
    constructor(uuid: string,
                docOwner: string,
                identity: (string | null),
                type: WaybillType,
                unitType: (UnitType | null),
                number: string,
                date: Date,
                shippingDate: Date,
                transport: (Transport | null),
                shipperId: string,
                consigneeId: string,
                supplierId: (string | null),
                base: (string | null),
                note: (string | null),
                status: WaybillStatus,
                resolution: WaybillResolution,
                ttnInformF2RegUuid: (string | null),
                wbRegId: (string | null),
                direction: Direction,
                version: Version) {
        this.uuid = uuid;
        this.docOwner = docOwner;
        this.identity = identity;
        this.type = type;
        this.unitType = unitType;
        this.number = number;
        this.date = date;
        this.shippingDate = shippingDate;
        this.transport = transport;
        this.shipperId = shipperId;
        this.consigneeId = consigneeId;
        this.supplierId = supplierId;
        this.base = base;
        this.note = note;
        this.status = status;
        this.resolution = resolution;
        this.ttnInformF2RegUuid = ttnInformF2RegUuid;
        this.wbRegId = wbRegId;
        this.direction = direction;
        this.version = version;
    }
}

/**
 * @class module:waybill.WaybillPosition
 * @classdesc Позиция в товарно-транспортной накладной.
 * @param {string} uuid Уникальный идентификатор
 * @param {string} waybillUuid Уникальный идентификатор накладной
 * @param {?string} productIdentity Информация о продукции
 * @param {module:productInfo.ProductInfo} productInfo Информация о продукции
 * @param {?string} packId Идентификатор упаковки
 * @param {number} quantity Количество
 * @param {number} price Цена за единицу товара
 * @param {?string} party Номер партии
 * @param {?string} identity Идентификатор позиции внутри накладной
 * @param {?string} informF1RegId Справка А
 * @param {?string} informF2RegId Регистрационный номер раздела справки Б
 */
export class WaybillPosition {
    constructor(uuid: string,
                waybillUuid: string,
                productIdentity: (string | null),
                productInfo: ProductInfo,
                packId: (string | null),
                quantity: number,
                price: number,
                party: (string | null),
                identity: (string | null),
                informF1RegId: (string | null),
                informF2RegId: (string | null)) {
        this.uuid = uuid;
        this.waybillUuid = waybillUuid;
        this.productIdentity = productIdentity;
        this.productInfo = productInfo;
        this.packId = packId;
        this.quantity = quantity;
        this.price = price;
        this.party = party;
        this.identity = identity;
        this.informF1RegId = informF1RegId;
        this.informF2RegId = informF2RegId;
    }
}

/**
 * @class module:waybill.WaybillPositionMark
 * @classdesc Марка в позиции товарно-транспортной накладной.
 * @param {string} waybillPositionUuid Уникальный идентификатор позиции в ТТН
 * @param {?string} boxNumber Номер короба
 * @param {string} mark Марка
 */
export class WaybillPositionMark {
    constructor(waybillPositionUuid: string,
                boxNumber: (string | null),
                mark: string) {
        this.waybillPositionUuid = waybillPositionUuid;
        this.boxNumber = boxNumber;
        this.mark = mark;
    }
}

File: android/src/main/java/com/evotor/egais/mapper/Waybill.kt
package com.evotor.egais.mapper

import android.database.Cursor
import com.evotor.egais.converter.MoneyBigDecimalConverter
import com.evotor.egais.converter.QuantityBigDecimalConverter
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

private fun createTransport(cursor: Cursor): WritableMap? {
    val transportType = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_TYPE"))
    val transportCompany = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_COMPANY"))
    val transportCar = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_CAR"))
    val transportTrailer = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_TRAILER"))
    val transportCustomer = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_CUSTOMER"))
    val transportDriver = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_DRIVER"))
    val transportLoadPoint = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_LOADPOINT"))
    val transportUnloadPoint = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_UNLOADPOINT"))
    val transportRedirect = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_REDIRECT"))
    val transportForwarder = cursor.getString(cursor.getColumnIndexOrThrow("TRANSPORT_FORWARDER"))
    if (transportType != null ||
            transportCompany != null ||
            transportCar != null ||
            transportTrailer != null ||
            transportCustomer != null ||
            transportDriver != null ||
            transportLoadPoint != null ||
            transportUnloadPoint != null ||
            transportRedirect != null ||
            transportForwarder != null) {
        val result = Arguments.createMap()
        result.putString("type", transportType)
        result.putString("company", transportCompany)
        result.putString("car", transportCar)
        result.putString("trailer", transportTrailer)
        result.putString("customer", transportCustomer)
        result.putString("driver", transportDriver)
        result.putString("loadpoint", transportLoadPoint)
        result.putString("unloadpoint", transportUnloadPoint)
        result.putString("redirect", transportRedirect)
        result.putString("forwarder", transportForwarder)
    }
    return null
}


internal fun createWaybill(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("docOwner", cursor.getString(cursor.getColumnIndexOrThrow("OWNER")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putString("type", cursor.getString(cursor.getColumnIndexOrThrow("TYPE")))
    result.putString("unitType", cursor.getString(cursor.getColumnIndexOrThrow("UNIT_TYPE")))
    result.putString("number", cursor.getString(cursor.getColumnIndexOrThrow("NUMBER")))
    result.putString("date", cursor.getString(cursor.getColumnIndexOrThrow("DATE")))
    result.putString("shippingDate", cursor.getString(cursor.getColumnIndexOrThrow("SHIPPING_DATE")))
    result.putMap("transport", createTransport(cursor))
    result.putString("shipperId", cursor.getString(cursor.getColumnIndexOrThrow("SHIPPER_ID")))
    result.putString("consigneeId", cursor.getString(cursor.getColumnIndexOrThrow("CONSIGNEE_ID")))
    result.putString("supplierId", cursor.getString(cursor.getColumnIndexOrThrow("SUPPLIER_ID")))
    result.putString("base", cursor.getString(cursor.getColumnIndexOrThrow("BASE")))
    result.putString("note", cursor.getString(cursor.getColumnIndexOrThrow("NOTE")))
    result.putString("status", cursor.getString(cursor.getColumnIndexOrThrow("STATUS")))
    result.putString("resolution", cursor.getString(cursor.getColumnIndexOrThrow("RESOLUTION")))
    result.putString("ttnInformF2RegUuid", cursor.getString(cursor.getColumnIndexOrThrow("TTN_INFORM_F2_REG_UUID")))
    result.putString("wbRegId", cursor.getString(cursor.getColumnIndexOrThrow("WB_REG_ID")))
    result.putString("direction", cursor.getString(cursor.getColumnIndexOrThrow("DIRECTION")))
    result.putString("version", cursor.getString(cursor.getColumnIndexOrThrow("VERSION")))
    return result
}

internal fun createWaybillPosition(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("waybillUuid", cursor.getString(cursor.getColumnIndexOrThrow("WAY_BILL_UUID")))
    result.putString("productIdentity", cursor.getString(cursor.getColumnIndexOrThrow("PRODUCT_INFO_IDENTITY")))
    result.putMap("productInfo", createProductInfo(cursor))
    result.putString("packId", cursor.getString(cursor.getColumnIndexOrThrow("PACK_ID")))
    result.putDouble("quantity", QuantityBigDecimalConverter.toBigDecimal(cursor.getLong(cursor.getColumnIndexOrThrow("QUANTITY"))).toDouble())
    result.putDouble("price", MoneyBigDecimalConverter.toBigDecimal(cursor.getLong(cursor.getColumnIndexOrThrow("PRICE"))).toDouble())
    result.putString("party", cursor.getString(cursor.getColumnIndexOrThrow("PARTY")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putString("informF1RegId", cursor.getString(cursor.getColumnIndexOrThrow("INFORM_F1_REG_ID")))
    result.putString("informF2RegId", cursor.getString(cursor.getColumnIndexOrThrow("INFORM_F2_REG_ID")))
    return result
}

internal fun createTtnInformF2Reg(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("docOwner", cursor.getString(cursor.getColumnIndexOrThrow("OWNER")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putString("wbRegId", cursor.getString(cursor.getColumnIndexOrThrow("WB_REG_ID")))
    result.putString("egaisFixNumber", cursor.getString(cursor.getColumnIndexOrThrow("EGAIS_FIX_NUMBER")))
    result.putString("egaisFixDate", cursor.getString(cursor.getColumnIndexOrThrow("EGAIS_FIX_DATE")))
    result.putString("wbNumber", cursor.getString(cursor.getColumnIndexOrThrow("WB_NUMBER")))
    result.putString("wbDate", cursor.getString(cursor.getColumnIndexOrThrow("WB_DATE")))
    result.putString("shipperId", cursor.getString(cursor.getColumnIndexOrThrow("SHIPPER_ID")))
    result.putString("consigneeId", cursor.getString(cursor.getColumnIndexOrThrow("CONSIGNEE_ID")))
    result.putString("supplierId", cursor.getString(cursor.getColumnIndexOrThrow("SUPPLIER_ID")))
    result.putString("waybillId", cursor.getString(cursor.getColumnIndexOrThrow("WAY_BILL_ID")))
    result.putString("status", cursor.getString(cursor.getColumnIndexOrThrow("STATUS")))
    return result
}

internal fun createTtnInformF2RegPosition(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("uuid", cursor.getString(cursor.getColumnIndexOrThrow("UUID")))
    result.putString("ttnInformF2RegUuid", cursor.getString(cursor.getColumnIndexOrThrow("TTN_INFORM_F2_REG_ID")))
    result.putString("identity", cursor.getString(cursor.getColumnIndexOrThrow("IDENTITY")))
    result.putString("informF2RegId", cursor.getString(cursor.getColumnIndexOrThrow("INFORM_F2_REG_ID")))
    return result
}

internal fun createWaybillPositionMark(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("waybillPositionUuid", cursor.getString(cursor.getColumnIndexOrThrow("WAYBILL_POSITION_UUID")))
    result.putString("boxNumber", cursor.getString(cursor.getColumnIndexOrThrow("BOX_NUMBER")))
    result.putString("mark", cursor.getString(cursor.getColumnIndexOrThrow("MARK")))
    return result
}

internal fun createWaybillActPositionMark(cursor: Cursor): WritableMap {
    val result = Arguments.createMap()
    result.putString("waybillActPositionUuid", cursor.getString(cursor.getColumnIndexOrThrow("WAY_BILL_ACT_POSITION_UUID")))
    result.putString("mark", cursor.getString(cursor.getColumnIndexOrThrow("MARK")))
    return result
}